Bug 706809 - write through cache for the sauropod backend (draft)

tarekziade opened the pull request:
We maintain the meta document in memcached to avoid hitting Sauropod on most GETs

billrest committed:
first draft of write-through cache for the meta document

rfk commented on appsync/storage/sauropod.py:
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
+
+    def _get_cached_metadata(self, session, user, collection):
+        # getting the cached value if possible
+        if self.cache is None:
```
Should be "is not None"?

rfk commented on appsync/storage/sauropod.py:
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
+        session.set(key, json.dumps(data), if_match=etag)
+
+        if self.cache:
+            doc = session.getitem(key)
```
the call to session.set() above should return the updated item, so you can avoid fetching it against here

tarekziade replied:
Oh, I just realized set() returns the Item. So yeah, changing this

rfk commented:
@dclarke: the _get_cached_metadata method calls through to hbase itself, so I guess it should just grow some logic to ignore e.g. connection errors from the cache.

rfk commented on appsync/storage/sauropod.py:
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
+        if self.cache is None:
+            self.cache.set(cache_key, doc, time=self.cache_ttl)
+
+        return res
```
res undefined, should be "return doc"?

tarekziade commented:
@dclarke this code has already been merged......

Not this is not merged, it's a ongoing work, a draft in my clone

dclarke commented:
rfk, looks correct

dclarke commented:
tarek: yup correct , i read incorrectly, looks good though

billrest committed:
Fixed the various issues found by Ryan.

Also: instead of letting the CacheError bubbling up and generate
a log + 500, we just generate a log and ignore it, so the
backend still works when Memcache is down.

billrest committed:
code cleanup

tarekziade commented:
@dclarke Adding a hit counter test now

billrest committed:
In this changeset, I am adding a test_sauropod module.

The test runs a fake Sauropod server and runs test_server
against it. This makes sure the sauropod backend is tested
even if there's no Node.js running.

The test also exercise the memcache layer.

Tests should pass if memcache is OFF or ON. OFF ignores
the cache and go through.

billrest committed:
code cleanup

billrest committed:
Now testing the number of hits to make sure Memcache really caches.

If Memcache is running the tests will use it and make sure it caches
Sauropod GETs on the meta document, and saves hits.

If Memcache is not running, the tests makes sure the backend still
works with no cache.

For the SQL backend, Memcache is mandatory for the tokens, if not
running a verify will generate a 503 and another call a 401.

tarekziade commented:
@dclarke in this last commit, I am counting the hits. 
- If Memcache is running the tests will use it and make sure it caches Sauropod GETs on the meta document, and saves hits.
- If Memcache is not running, the tests makes sure the backend still works with no cache.
- For the SQL backend, Memcache is mandatory for the tokens, if not running a verify will generate a 503 and another call a 401.

I think the change is ready to be commited now, let me know guys.

billrest committed:
forgot to add this file

rfk commented on appsync/storage/sauropod.py:
```diff
@@ -132,13 +148,63 @@ def _resume_session(self, token):
+            try:
+                self.cache.set(cache_key, doc, self.cache_ttl)
+            except CacheError:
+                logger.error('Unable to read the metadata in the cache.')
```
s/read/write/

billrest committed:
typo
